Direction tagging and emptyFrom

johnomotani opened the pull request:
Adds members to `Field` to tag the type of grid directions. In particular, this is useful to tag fields as either `YOrthogonal` (i.e. defined on a grid where y and z directions are orthogonal, as for `FCITransform` and the default case for `ShiftedMetric`) or `YAligned` (i.e. defined on a field-aligned grid where y and z directions are not orthogonal, as for `ParallelTransformIdentity` or fields which have been transformed with `toFieldAligned` by `ShiftedMetric`).

Expands constructors for `Field`, `Field3D`, `Field2D` and `FieldPerp` with extra arguments to set the new members, and also `location`. Also adds a new `T emptyFrom(const T&f)` function to create a new field with meta-data copied from an existing one, and data allocated but not initialized. Using `emptyFrom` means the new long constructors are hardly ever needed explicity. Modifies the 'setter' methods to return a reference to the field, allowing method chaining if something needs to be changed, for example `Field3D result{emptyFrom(f).setLocation(outloc)};`.

There are a fair few files that are only changed by using `emptyFrom` for new fields and `fieldsCompatible` for assert checks. Here's a link to compare up to e195604, before I added most of those: https://github.com/boutproject/BOUT-dev/compare/next...e195604babcf0fac9897ff413e0f611f82e3d34b

Todo list (feel free to add to this):
- [ ] add some explanation about `YOrthogonal`/`YAligned`, `emptyFrom` and `fieldsCompatible` to the manual

ZedThree commented on include/fieldperp.hxx:
```diff
@@ -254,16 +257,13 @@ class FieldPerp : public Field {
 private:
   /// The Y index at which this FieldPerp is defined
-  int yindex{-1};
+  int yindex;
```
Why the removal of the default value?

johnomotani replied:
That was a mistake. I'll fix it.

ZedThree commented on src/mesh/interpolation/bilinear.cxx:
```diff
@@ -28,7 +28,11 @@
+  w0(localmesh, CELL_CENTRE, DIRECTION::X, DIRECTION::YOrthogonal, DIRECTION::Z),
+  w1(localmesh, CELL_CENTRE, DIRECTION::X, DIRECTION::YOrthogonal, DIRECTION::Z),
+  w2(localmesh, CELL_CENTRE, DIRECTION::X, DIRECTION::YOrthogonal, DIRECTION::Z),
+  w3(localmesh, CELL_CENTRE, DIRECTION::X, DIRECTION::YOrthogonal, DIRECTION::Z) {
```
Do these need explicit values?

johnomotani replied:
In the present implementation, we have to set explicit values here, because these fields need to be allocated before the `ParallelTransform` has finished being constructed, so we can't get the default `yDirectionType` from the `ParallelTransform` without creating a circular call. Because this class is only used by `FCITransform` we know what the default should be, so we can set it explicitly.

ZedThree commented on src/mesh/interpolation/hermite_spline.cxx:
```diff
@@ -28,9 +28,15 @@
+      h00_z(localmesh, CELL_CENTRE, DIRECTION::X, DIRECTION::YOrthogonal, DIRECTION::Z),
+      h01_z(localmesh, CELL_CENTRE, DIRECTION::X, DIRECTION::YOrthogonal, DIRECTION::Z),
+      h10_z(localmesh, CELL_CENTRE, DIRECTION::X, DIRECTION::YOrthogonal, DIRECTION::Z),
+      h11_z(localmesh, CELL_CENTRE, DIRECTION::X, DIRECTION::YOrthogonal, DIRECTION::Z) {
```
Similarly, can these just use the defaults? I'm sure there's a reason you've done this, but it's not obvious to me, sorry!

ZedThree commented on src/mesh/parallel/shiftedmetric.cxx:
```diff
@@ -136,18 +136,47 @@ void ShiftedMetric::cachePhases() {
+    return f;
+  default:
+#if CHECK > 0
+    throw BoutException("Unrecognized y-direction type for Field3D passed to "
```
Can we just always do this? This is essentially free, so I don't see a downside to removing the `#if CHECK`

ZedThree commented on src/mesh/parallel/shiftedmetric.cxx:
```diff
@@ -136,18 +136,47 @@ void ShiftedMetric::cachePhases() {
+    // f is already in orthogonal coordinates
+    return f;
+  default:
+#if CHECK > 0
```
See above -- this check is free, let's always do it

ZedThree commented on src/sys/bout_types.cxx:
```diff
@@ -73,3 +81,88 @@ const std::string& DERIV_STRING(DERIV deriv) {
+  }
+}
+
+bool compatibleDirections(DIRECTION d1, DIRECTION d2) {
```
How about the following?

```cpp
return (isXDirectionType(d1) && isXDirectionType(d2))
    || (isYDirectionType(d1) && isYDirectionType(d2))
    || (isZDirectionType(d1) && isZDirectionType(d2))
    || (isSpecialDirectionType(d1) || isSpecialDirectionType(d1));
```

(needs `isSpecialDirectionType` to be implemented)

johnomotani replied:
Almost, need to add another condition so that we return `false` if `d1==YAligned` and `d2==YOrthogonal` or vice versa. Still think this looks nicer than the way I have it at the moment.

johnomotani replied:
actually, I'm not so sure about that. I want `YOrthogonal` to be compatible with `YOrthogonal` or `Y` but not `YAligned`, and `YAligned` to be compatible with `YAligned` or `Y` but not `YOrthogonal`. The single conditional gets a bit convoluted. Do you think it's still nicer than the nested switch-case statements?

ZedThree commented on src/sys/derivs.cxx:
```diff
@@ -229,8 +229,8 @@ const Field3D D2DZ2(const Field3D &f, CELL_LOC outloc, const std::string &method
-  auto tmp = Field2D(0., f.getMesh());
-  tmp.setLocation(f.getLocation());
+  Field2D tmp{emptyFrom(f)};
+  tmp = 0.;
```
Could just do `return (emptyFrom(f) = 0.)`, but it does look a bit weird

ZedThree commented on tests/unit/mesh/parallel/test_shiftedmetric.cxx:
```diff
@@ -37,6 +37,14 @@ class ShiftedMetricTest : public ::testing::Test {
+        Field2D{0.0}, Field2D{1.0}, Field2D{1.0}, Field2D{1.0}, Field2D{0.0},
+        Field2D{0.0}, Field2D{0.0}, Field2D{0.0}, Field2D{0.0}, false));
+
+    mesh->setParallelTransform(Mesh::PTptr(new ShiftedMetric(*mesh, zShift)));
```
Use `bout::utils::make_unique<ShiftedMetric>(*mesh, zShift)` here

ZedThree commented on tests/unit/test_extras.hxx:
```diff
@@ -149,6 +149,7 @@ public:
     maxregionblocksize = MAXREGIONBLOCKSIZE;
 
     setCoordinates(nullptr);
+    setParallelTransform(PTptr(new ParallelTransformIdentity(*this)));
```
`make_unique`

ZedThree commented on include/bout/paralleltransform.hxx:
```diff
@@ -92,6 +94,11 @@ public:
   }
+
+  DIRECTION getDefaultYDirectionType() const override {
+    return DIRECTION::YAligned;
```
I'm not sure about this at all. This says that the identity transform is synonymous with a field-aligned grid, and I'm not convinced that's true.

What does saying the identity transform has field-aligned fields by default get us?

ZedThree commented:
Lots of comments, sorry! This is definitely the direction we want to be going in. There's a few things we should iron out first though.

I'm quite wary of `ParallelTransform::getDefaultYDirectionType`. It adds some complications in some places, and I can't see the benefit. What does it get us, and is that worth the complexity it introduces?

johnomotani commented on include/bout/empty_from.hxx:
```diff
@@ -0,0 +1,21 @@
+/// Return an empty shell field of some type derived from Field, with metadata
+/// copied but empty data array
+template<typename T>
+inline T emptyFrom(const T& f) {
```
That's probably better. The reason for having a separate file was that I wanted to put the general version and the `FieldPerp` specialization in the same place; I couldn't include `fieldperp.hxx` in `field.hxx` because of the circular dependency, so put them both in a new file. In hindsight it would be better to have `emptyFrom` in `field.hxx` and the `FieldPerp` specialization in `fieldperp.hxx`.

johnomotani commented on include/field.hxx:
```diff
@@ -95,6 +114,11 @@ class Field {
+      // allocate() is called in one of the derived classes. fieldmesh==nullptr
+      // indicates that some initialization that would be done in the
+      // constructor if fieldmesh was a valid Mesh object still needs to be
+      // done.
```
I agree this hasn't ended up being a good implementation, but I don't think we can just return `nullptr` here:
* If `fieldmesh==nullptr` then when the field is allocated, we will set `fieldmesh=bout::globals::mesh`, and since `getMesh()` might be called before the field is allocated, we want to return `bout::globals::mesh`.
* The reason we can't set it here is that we use `fieldmesh==nullptr` to indicate that `allocate()` needs to do some additional set-up that couldn't be done in the constructor because we didn't have a valid mesh then: namely setting, if needed for the child class, `nx`/`ny`/`nz` and default values of `xDirectionType`/`yDirectionType`/`zDirectionType`.

Would be nice for this to be able to require fields to have a valid mesh in the constructor, but that would break the interface too much.

Would it be better to add something like `bool fullyInitialised{false}` to `Field` and check/update that in `allocate()`? Then the intention would be clearer from the variable name and we could set `fieldmesh = bout::globals::mesh` in `getMesh()` too.

ZedThree replied:
Ok, looks like we need to do some more thinking about what we should/want to be doing here. Definitely future work though

johnomotani commented on include/interpolation.hxx:
```diff
@@ -62,138 +62,140 @@ const T interp_to(const T& var, CELL_LOC loc, REGION region = RGN_ALL) {
+  if (var.getLocation() == loc) {
+    // Nothing to do - just return unchanged
+    // Copying into result to return as returning var may increase the number of
+    // references to the var data whilst returning result doesn't
```
This was a suggestion from @d7919 in https://github.com/boutproject/BOUT-dev/pull/921#discussion_r172671112:
> I think we may still want to copy var into result. As it stands returning var may increase the number of references to the var data whilst returning result doesn't. This may or may not be an important difference.

d7919 replied:
(I may have been very wrong)

ZedThree replied:
I think if we did `result.ensureUnique` before returning it, that would be the case, but also a pessimisation as we would have to copy the actual data. We should be able to safely `return var`.

johnomotani commented on src/field/field.cxx:
```diff
@@ -32,14 +32,58 @@
+  // Need to check for nullptr again, because the fieldmesh might still be
+  // nullptr if the global mesh hasn't been initialized yet
+  if (fieldmesh != nullptr) {
+    // get Coordinates for our location from fieldmesh
```
I think it needed a better comment, like `// sets fieldCoordinates by getting Coordinates for our location from fieldmesh`. Since `getCoordinates()` is for getting `fieldCoordinates` *from* the field, this is (to me) a non-obvious use, in that the fact that `getCoordinates()` *sets* `fieldCoordinates` is a sort of side-effect. I'll update the comment.

johnomotani commented:
There's a few points that I think tie together:
* I think `ParallelTransformIdentity` means a field-aligned grid/coordinate system, because it says we can do `Grad_par` as a y-derivative without any transformation, and we don't necessarily have to have orthogonal x-z coordinates in a simulation using `ParallelTransformIdentity` (e.g. sheared slab). That's what motivated me to make `YAligned` the default for `ParallelTransformIdentity`.
    * practically, it would be just as good at the moment to have `YOrthogonal` the default for `ParallelTransformIdentity` because it doesn't change the y-derivatives for `ParallelTransformIdentity`. As @ZedThree says, having `YOrthogonal` the default for everything would simplify things a fair bit.
* The reason for having `DIRECTION::Null` and using it for defaults is to support having default direction types that could depend on the type of `ParallelTransform`, `Mesh`, `Coordinates`, etc. I still think this is a potentially useful feature, but it's true we could do without it for the moment.

So at the moment we don't need the feature to set the default `yDirectionType`, we could just use `YOrthogonal`. I guess we're not likely to add a new `ParallelTransform` any time soon, that might require a different default. I can imagine it would be more likely that we could add some functionality that required x-z to be orthogonal (`g13==0`), in which case we might want to set `YAligned` as the default for `ParallelTransformIdentity` so we could add checks assuming that `YOrthogonal` really means x-z are orthogonal.

What do people prefer?

johnomotani commented on include/field3d.hxx:
```diff
@@ -167,7 +168,10 @@ class Field3D : public Field, public FieldData {
+  Field3D(Mesh *localmesh = nullptr, CELL_LOC location_in=CELL_CENTRE,
+          DIRECTION xDirectionType_in=DIRECTION::Null,
+          DIRECTION yDirectionType_in=DIRECTION::Null,
+          DIRECTION zDirectionType_in=DIRECTION::Null);
```
The reason is to ensure that the `fieldsCompatible` check would fail if setNullDirectionTypesToDefault() hasn't been called, which it should have been if the field has been allocated. Using `X`, `Y`, `Z` for defaults this wouldn't be true because `Y` is compatible with `Y`, `YOrthogonal` or `YAligned.

johnomotani commented on src/field/fieldperp.cxx:
```diff
@@ -378,12 +338,11 @@ const FieldPerp sliceXZ(const Field3D& f, int y) {
 
-  FieldPerp result(f.getMesh());
+  FieldPerp result(f.getMesh(), f.getLocation(), y, f.getDirectionX(),
+                   f.getDirectionY(), f.getDirectionZ());
```
Here we can't use `emptyFrom` because `f` is a `Field3D` while `result` is a `FieldPerp`. This only happens in a couple of places, so it didn't seem worth adding a template or set of functions to handle it.

ZedThree commented on src/mesh/parallel/shiftedmetric.cxx:
```diff
@@ -136,18 +136,41 @@ void ShiftedMetric::cachePhases() {
-  return shiftZ(f, toAlignedPhs, region);
+  switch (f.getDirectionY()) {
+  case (DIRECTION::YOrthogonal):
+    return shiftZ(f, toAlignedPhs, DIRECTION::YOrthogonal, region);
```
This should use `DIRECTION::YAligned`

johnomotani replied:
Good spot! Guess this shows I should add more unit tests...

ZedThree commented:
> I think `ParallelTransformIdentity` means a field-aligned grid/coordinate system, because it says we can do Grad_par as a y-derivative without any transformation

But simulations without magnetic fields aren't field-aligned, so I don't think that's quite right. The identity transform just means "don't do a parallel transform for parallel derivatives", which isn't quite the same as saying that the grid is field-aligned.
I'd really rather as much of the code as possible doesn't care about what transform is being used.

A few more points I've been thinking about:
- Is it possible to drop `DIRECTION::YOrthogonal` completely? We put it in originally in order to handle derivatives using the parallel slices slightly differently. As far as I can tell, the only place it's actually _used_ is in `populateStencil`, where it could be replaced with `f.hasYUpDown()`?
- Possibly separately from the above, does it make more sense to move `Special` and `YAligned` out of `DIRECTION` into separate enums? We might then have something like `enum class YDirectionType {Full, Slice, Aligned}` and `enum class ZDirectionType {Full, Average}` (or perhaps combine them into a single enum). It currently feels a bit redundant to say that the x-direction-type is x.

ZedThree commented on include/bout_types.hxx:
```diff
@@ -49,10 +49,12 @@ enum REGION {RGN_ALL, RGN_NOBNDRY, RGN_NOX, RGN_NOY, RGN_NOZ};
 
 /// To identify particular directions (in index space)
-enum class DIRECTION { X = 0, Y = 1, Z = 3, YAligned = 4, YOrthogonal = 5 };
+enum class DIRECTION { X = 0, Y = 1, Z = 3, YAligned = 4, YOrthogonal = 5, Special = 6, Null = 7 };
```
Also I realise you've only updated this bit, but note that we don't need the explicit initialisers here, we could just have `enum class DIRECTION {X, Y, Z, ...}`

johnomotani replied:
Sounds right to me. @d7919 was there a reason for skipping `2` when initializing the values?

d7919 replied:
I don't remember I'm afraid -- I suspect not. Probably I had something like `{X, Y, YOther, Z}` and then wanted to expand to more `Y` types and was easiest to just stick both "special" cases at the end rather than leave `YOther` where it was and add another at the end.

d7919 commented:
> > I think `ParallelTransformIdentity` means a field-aligned grid/coordinate system, because it says we can do Grad_par as a y-derivative without any transformation
> 
> But simulations without magnetic fields aren't field-aligned, so I don't think that's quite right. The identity transform just means "don't do a parallel transform for parallel derivatives", which isn't quite the same as saying that the grid is field-aligned.
> I'd really rather as much of the code as possible doesn't care about what transform is being used.
> 
> A few more points I've been thinking about:
> 
> * Is it possible to drop `DIRECTION::YOrthogonal` completely? We put it in originally in order to handle derivatives using the parallel slices slightly differently. As far as I can tell, the only place it's actually _used_ is in `populateStencil`, where it could be replaced with `f.hasYUpDown()`?
> * Possibly separately from the above, does it make more sense to move `Special` and `YAligned` out of `DIRECTION` into separate enums? We might then have something like `enum class YDirectionType {Full, Slice, Aligned}` and `enum class ZDirectionType {Full, Average}` (or perhaps combine them into a single enum). It currently feels a bit redundant to say that the x-direction-type is x.

I think the original idea of not using `hasYupYdown` in populate stencil was to avoid introducing a runtime branch into this code as it's right at the bottom of the nested loops -- using the enum class as a template parameter/known at compile time for each call should allow the branching to be determined at compile time (I think).

ZedThree commented:
> I think the original idea of not using `hasYupYdown` in populate stencil was to avoid introducing a runtime branch into this code as it's right at the bottom of the nested loops -- using the enum class as a template parameter/known at compile time for each call should allow the branching to be determined at compile time (I think).

Ah yes, good point! This is why we still need to create separate derivatives for `Y` and `YOrthogonal`. Ok, we probably do need to keep it then.

johnomotani commented:
> But simulations without magnetic fields aren't field-aligned, so I don't think that's quite right. The identity transform just means "don't do a parallel transform for parallel derivatives", which isn't quite the same as saying that the grid is field-aligned.

I'd not considered simulations with no magnetic field. Even then though, isn't the grid's y-direction in some sense aligned to whatever `Grad_par` is parallel to (if using the identity transform)?

> I'd really rather as much of the code as possible doesn't care about what transform is being used.

Agree with this. Would it be better to say that regardless of which transform we're using fields all start out just in default state (i.e. `yDirectionType = Y`), and then change the `yDirectionType` if we do something special like transforming to field-aligned?

I agree it would make sense to make separate `enum class YDirectionType` and `ZDirectionType`. It would simplify things, and since it makes no sense to have `yDirectionType = DIRECTION::X` the general `DIRECTION` doesn't really fit for what we've used it for here.

johnomotani commented:
@ZedThree @d7919 do you think it's useful to have a `YDirectionType::Slice`? I guess it would indicate that a `Field3D` has parallel slices, so we could use it to replace `hasYUpDown()`. I don't see an obvious benefit of doing that over leaving `yDirectionType=YDirectionType::Y` and using `hasYUpDown()` as we do at the moment.

johnomotani commented:
I've had a go at implementing some of what we were just discussing. There are now `YDirectionType` and `ZDirectionType` enum classes, and they get sensible defaults `*DirectionType::Standard`*, independent of the `ParallelTransform`. I also decided to store the direction types in `DirectionTypes` struct, because checking compatibility of directions now needs both direction types: `YDirectionType::Aligned` is compatible with `YDirectionType::Standard` if the `Standard` field has `ZDirectionType::Average`.

@ZedThree I think this addresses most of your remaining comments?

--
\* except `Field2D` which uses `ZDirectionType::Average`

ZedThree commented on include/bout_types.hxx:
```diff
@@ -48,21 +48,49 @@ enum REGION {RGN_ALL, RGN_NOBNDRY, RGN_NOX, RGN_NOY, RGN_NOZ};
+/// - Standard is the default for the Mesh/Coordinates/ParallelTransform
+/// - Aligned indicates that the field has been transformed to field-aligned
+///   coordinates
+enum class YDirectionType { Standard, Aligned };
```
Should this also include something like `Slice` for `FieldPerp`?

johnomotani replied:
`FieldPerp` is different depending on whether it's a slice of a standard `Field3D` or a field-aligned one. At the moment it just uses the same `YDirectionTypes` as the `Field3D`. I think that's the simplest way: we could have `Slice` and `SliceAligned` values too and add code to translate/compare those, but I can't think of anywhere we'd need to use or be able to check the difference between `Slice`/`Standard` or `SliceAligned`/`Aligned`. Do you have something in mind?

ZedThree replied:
Nothing in particular, just trying to think through everything

ZedThree commented on include/bout_types.hxx:
```diff
@@ -48,19 +48,49 @@ enum REGION {RGN_ALL, RGN_NOBNDRY, RGN_NOX, RGN_NOY, RGN_NOZ};
+
+/// Check whether direction types are compatible, so two fields with attributes
+/// d1 and d2 respectively can be added, subtracted, etc.
+bool compatibleDirections(const DirectionTypes& d1, const DirectionTypes& d2);
```
If I could bikeshed this a little, function names should be verbs where practical. So maybe `areDirectionsCompatible`?

ZedThree commented on include/field.hxx:
```diff
@@ -121,11 +140,65 @@ class Field {
+    swap(first.directions, second.directions);
+  }
+
+  friend bool fieldsCompatible(const Field& field1, const Field& field2) {
```
Does this need to be a friend? There are getters for the directions, so it would be cleaner to use those

ZedThree commented on include/field.hxx:
```diff
@@ -121,11 +140,65 @@ class Field {
+        // need to check fieldmesh and location, at least for now
+        field1.getMesh() == field2.getMesh() &&
+        field1.getLocation() == field2.getLocation() &&
+        // Compatible directions
```
I think the code is clear enough you can drop all these comments :)

ZedThree commented on include/field.hxx:
```diff
@@ -121,11 +140,65 @@ class Field {
 };
 
+/// Return an empty shell field of some type derived from Field, with metadata
+/// copied but empty data array
```
Please could you add something explicit here about the field being allocated but uninitialised?

ZedThree commented on src/field/field3d.cxx:
```diff
@@ -72,21 +74,18 @@ Field3D::Field3D(const Field3D& f) : Field(f.fieldmesh), data(f.data) {
 }
 
-Field3D::Field3D(const BoutReal val, Mesh* localmesh) : Field(localmesh) {
+Field3D::Field3D(const BoutReal val, Mesh* localmesh) : Field3D(localmesh) {
```
This now delegates to the `Field3D` ctor that sets `nx` etc. so we can probably remove L92-94 below that also set them

ZedThree commented on src/field/field3d.cxx:
```diff
@@ -143,8 +145,8 @@ void Field3D::splitYupYdown() {
-    yup_fields.emplace_back(fieldmesh);
-    ydown_fields.emplace_back(fieldmesh);
+    yup_fields.emplace_back(emptyFrom(*this));
+    ydown_fields.emplace_back(emptyFrom(*this));
```
Note that this now immediately allocates. I'm not sure we want to do this here as it should be done in the parallel transforms.

Although maybe it doesn't hurt as it shouldn't actually allocate memory in most cases.

johnomotani replied:
Still, it wouldn't hurt to just use the full constructor here to avoid the `allocate()`. It might save a bit of memory usage in some cases. I'll update.

ZedThree replied:
I'm not even sure we need the full constructor -- just some kind of object in the vector that we can assign to. The transforms don't (or shouldn't?) update the field, but clobber it completely.

ZedThree commented on src/field/field3d.cxx:
```diff
@@ -263,13 +233,16 @@ Field3D & Field3D::operator=(const Field3D &rhs) {
+  nz = rhs.nz;
 
-  setLocation(rhs.location);
+  ASSERT1(fieldsCompatible(*this, rhs));
```
I think this be done immediately after `TRACE` and before `copyFieldMembers`. It seems likely this test will always pass in this location

johnomotani replied:
If `copyFieldMembers` is working correctly then this test should indeed always pass. So it's only a sanity check that `copyFieldMembers` and `areFieldsCompatible` haven't somehow become inconsistent. It's probably better to check that in a unit test than here...

ZedThree commented on src/mesh/parallel/fci.cxx:
```diff
@@ -73,11 +73,16 @@ FCIMap::FCIMap(Mesh& mesh, int offset_, BoundaryRegionPar* boundary, bool zperio
   // Index-space coordinates of forward/backward points
-  Field3D xt_prime(&map_mesh), zt_prime(&map_mesh);
+  Field3D xt_prime(&map_mesh);
+  Field3D zt_prime{emptyFrom(xt_prime)};
```
Do we need `emptyFrom` here if we're just going to use `Mesh::get`? Unless that checks that the fields are compatible?

johnomotani replied:
Ah, this use of `emptyFrom` was only useful when I was calling the constructor for `xt_prime` with more arguments, that then wanted to be copied. It doesn't add anything any more, so will revert.

ZedThree commented on src/sys/bout_types.cxx:
```diff
@@ -52,6 +53,36 @@ const std::string& DIRECTION_STRING(DIRECTION direction) {
+  }
+
+  if (d2.z == ZDirectionType::Average && d2.y == YDirectionType::Standard
+      && (d1.y == YDirectionType::Standard || d1.y == YDirectionType::Aligned)) {
```
This second part will always be true. Can we just combine these two `if` statements?

johnomotani replied:
We could at the moment, but I think this way is stricter in case we ever add more `YDirectionType`s. I think actually I should check `d1.z == ZDirectionType::Standard` here as well, to be as strict as possible.

ZedThree commented on tests/unit/test_extras.hxx:
```diff
@@ -153,6 +153,7 @@ public:
     maxregionblocksize = MAXREGIONBLOCKSIZE;
 
     setCoordinates(nullptr);
+    setParallelTransform(bout::utils::make_unique<ParallelTransformIdentity>(*this));
```
This is undefined behaviour, btw, as `this` is not a complete `Mesh` object until _after_ the constructor has finished. If we do need to do this, it should be done in `FakeMeshFixture`

ZedThree commented:
Sorry @johnomotani, more comments!

ZedThree commented on include/bout/fv_ops.hxx:
```diff
@@ -194,7 +193,8 @@ namespace FV {
 
-    Field3D result = 0.0;
+    Field3D result{emptyFrom(f_in)};
+    result = 0.0;
```
We actually do this a fair bit, use `emptyFrom` and then set to 0.0. Is it worth having another helper function like `zeroFrom(f)`?

johnomotani replied:
How about another modifier method? So we could write something like `Field3D result{emptyFrom(f_in).fillZeros();}?

johnomotani replied:
...on the other hand, I think `zeroFrom` would take less code to implement, and require less typing to use. Seems like a good reason to add a new function instead of more methods :+1:

ZedThree replied:
Yeah, I think another free function is better than a method. It looks like we currently only need a `zeroFrom(Field)` rather than a more generic `fillFrom(BoutReal, Field)`?

johnomotani replied:
Yep, all the places we use `emptyFrom` we only ever set to zero afterwards, not to any non-zero value.

ZedThree commented:
@johnomotani Please could you also add unit tests for `emptyFrom` and any other new functions/methods? I think you have covered most of them already, but I didn't spot tests for that

ZedThree commented on include/bout/index_derivs_interface.hxx:
```diff
@@ -83,8 +83,7 @@ T flowDerivative(const T& vel, const T& f, CELL_LOC outloc, const std::string& m
   if (nPoint == 1) {
-    T tmp{emptyFrom(f).setLocation(outloc)};
-    tmp = 0.;
+    T tmp{zeroFrom(f).setLocation(outloc)};
```
You could go even further and just `return zeroFrom(f).setLocation(outloc)`

ZedThree commented on src/sys/derivs.cxx:
```diff
@@ -229,8 +229,7 @@ const Field3D D2DZ2(const Field3D &f, CELL_LOC outloc, const std::string &method
                     REGION UNUSED(region)) {
-  Field2D tmp{emptyFrom(f)};
-  tmp = 0.;
+  Field2D tmp{zeroFrom(f)};
```
`return zeroFrom(f);` :)

ZedThree commented on include/bout/fv_ops.hxx:
```diff
@@ -344,10 +343,10 @@ namespace FV {
   const Field3D Div_f_v(const Field3D &n_in, const Vector3D &v, bool bndry_flux) {
-    ASSERT2(n_in.getLocation() == v.getLocation());
+    ASSERT1(n_in.getLocation() == v.getLocation());
+    ASSERT1(areFieldsCompatible(n_in, v.x));
```
Do we also need the location check on `v` if we're checking `v.x`?

johnomotani replied:
It does: the location check on `v` makes sure its location isn't `CELL_VSHIFT`, which isn't supported by this function.

ZedThree commented:
:tada: Thanks @johnomotani ! A really great PR, I think `emptyFrom` especially leads to a lot cleaner code. Sorry for my million comments!

I suspect we need some experience actually using the direction tagging before being completely confident we've got exactly the right design, but I'm sure we're most of the way there already.

johnomotani commented:
Thanks @ZedThree. All the comments made the final code a lot simpler, more logical and cleaner!

d7919 commented on src/physics/smoothing.cxx:
```diff
@@ -45,7 +45,7 @@
   Mesh *mesh = f.getMesh();
-  Field3D result(mesh);
+  Field3D result{emptyFrom(f)};
   result.allocate();
```
This allocate should be removed

d7919 commented on src/sys/derivs.cxx:
```diff
@@ -329,18 +331,19 @@ const Field3D D2DXDZ(const Field3D &f, CELL_LOC outloc, const std::string &metho
-  Field3D result(f.getMesh());
+  Field3D result{emptyFrom(f)};
   ASSERT1(outloc == CELL_DEFAULT || outloc == f.getLocation());
   result.allocate();
```
Remove please.

d7919 commented on src/mesh/difops.cxx:
```diff
@@ -482,8 +457,7 @@ const Field2D Div_par_CtoL(const Field2D &var) {
 
-  Field3D result(mesh);
-  result.allocate();
+  Field3D result{emptyFrom(var).setLocation(CELL_CENTRE)};
```
This looks like the wrong location
